lib/puppet/application/apply.rb

```diff
@@ -185,17 +185,23 @@ def main
       facts.name = Puppet[:node_name_value]
     end
 
-    configured_environment = Puppet.lookup(:current_environment)
+    # Find our Node
+    unless node = Puppet::Node.indirection.find(Puppet[:node_name_value])
+      raise "Could not find node #{Puppet[:node_name_value]}"
+    end
+
+    configured_environment = node.environment || Puppet.lookup(:current_environment)
+
     apply_environment = manifest ?
       configured_environment.override_with(:manifest => manifest) :
       configured_environment
 
-    Puppet.override({:current_environment => apply_environment}, "For puppet apply") do
-      # Find our Node
-      unless node = Puppet::Node.indirection.find(Puppet[:node_name_value])
-        raise "Could not find node #{Puppet[:node_name_value]}"
-      end
```

Comment:
Previously, we looked up our node within the context of the `apply_environment`. So if the client had `environment=foo`, then the node request would have been `/foo/node/<name>`. Now, we retrieve the node definition without changing the environment context, but that should be ok, because we'll continue to use the current environment name in the request?